### PR TITLE
Sample connection profiles for S3 and OpenID Connect Federation

### LINKED ISSUE #55: Sample connection profiles for S3 and OpenID Connect Federation
> With web identity federation, you don't need to create custom sign-in code or manage your own user identities. Instead, users of your app can sign in using a well-known external identity provider (IdP), such as Login with Amazon, Facebook, Google, or any other [OpenID Connect (OIDC)](http://openid.net/connect/)-compatible IdP. They can receive an authentication token, and then exchange that token for temporary security credentials in AWS that map to an IAM role with permissions to use the resources in your AWS account. [^1] 

[^1]: https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_oidc.html

> The role that your application assumes must trust the identity provider that is associated with the identity token. In other words, the identity provider must be specified in the role's trust policy. The call to `AssumeRoleWithWebIdentity` should include the ARN of the role that is specific to the provider through which the user signed in.

→ https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_request.html#api_assumerolewithwebidentity

With support to retrieve authentication tokens for S3 using the STS APIs [^2] we should provide (sample) connection profiles for typical providers supported.

- [x] iterate-ch/cyberduck#14972

[^2]: iterate-ch/cyberduck#13804

# General [connection profile](https://docs.cyberduck.io/protocols/profiles/) blueprint

```
<?xml version="1.0" encoding="UTF-8"?>
<!--
  ~ Copyright (c) 2002-2023 iterate GmbH. All rights reserved.
  ~ https://cyberduck.io/
  ~
  ~ This program is free software; you can redistribute it and/or modify
  ~ it under the terms of the GNU General Public License as published by
  ~ the Free Software Foundation, either version 3 of the License, or
  ~ (at your option) any later version.
  ~
  ~ This program is distributed in the hope that it will be useful,
  ~ but WITHOUT ANY WARRANTY; without even the implied warranty of
  ~ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  ~ GNU General Public License for more details.
  -->

<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
    <dict>
        <key>Protocol</key>
        <string>s3</string>
        <key>Vendor</key>
        <string>s3-sts</string>
        <key>OAuth Authorization Url</key>
        <string>…</string>
        <key>OAuth Token Url</key>
        <string>…</string>
        <key>OAuth Client ID</key>
        <string>…</string>
        <key>OAuth Client Secret</key>
        <string>…</string>
        <key>OAuth Redirect Url</key>
        <string>x-cyberduck-action:oauth</string>
        <key>OAuth PKCE</key>
        <false/>
        <key>Scopes</key>
        <array>
            <string>openid</string>
            <string>offline_access</string>
        </array>
        <key>Password Configurable</key>
        <false/>
        <key>Username Configurable</key>
        <false/>
        <key>Token Configurable</key>
        <false/>
        <key>Username Placeholder</key>
        <string>Username</string>
        <key>STS Endpoint</key>
        <string>https://sts.amazonaws.com/</string>
        <key>Properties</key>
        <array>
            <string>s3.assumerole.rolearn=arn:aws:iam::…</string>
        </array>
    </dict>
</plist>
```

> Set `s3.assumerole.rolearn=` to an empty value for a prompt on login to enter it manually by the user.

## 🟥 AWS S3/STS + GitHub (OIDC) [^3] [^4] 

[^3]: https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-idp_oidc.html#idp_oidc_Create_GitHub
[^4]: https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-amazon-web-services

Prerequisites

- [ ] Configure OAuth App in GitHub [^6]
- [ ] Add the GitHub OIDC provider `https://token.actions.githubusercontent.com` to AWS IAM. For the _Audience_: Use `sts.amazonaws.com` [^7] 
- [ ] [Configure the role and trust policy](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-amazon-web-services#configuring-the-role-and-trust-policy) in AWS IAM

[^6]: https://github.com/organizations/iterate-ch/settings/applications/2282978
[^7]: https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-amazon-web-services#adding-the-identity-provider-to-aws

❗ Currently not possible to integrate because the id-token endpoint does only seem to be accessible from [GitHub actions](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-amazon-web-services#updating-your-github-actions-workflow).

## 🟢 AWS S3/STS + Google (OIDC) [^5] 

[^5]: https://cloud.google.com/storage-transfer/docs/source-amazon-s3#federated_identity 

- [x] Register OAuth Client ID in Google Cloud Console or use the registration from Cyberduck connection profiles for [Google Cloud Storage](https://github.com/iterate-ch/cyberduck/blob/master/profiles/default/Google%20Cloud%20Storage.cyberduckprofile) and [Google Drive](https://github.com/iterate-ch/cyberduck/blob/master/profiles/default/Google%20Drive.cyberduckprofile).
- [ ] Add the Google OIDC provider to AWS IAM. The _Audience_ must be set to the OAuth Client ID like `996125414232-s922bvdt21nceeh5dq1gb6av8plpj7hr.apps.googleusercontent.com`.
- [ ] Make sure to restrict access by configuring the role and trust policy

In the sample configuration the custom trust policy set for the IAM role in AWS, the condition limits access to a particular Google Account using 

```
…
   "Condition": {
        "StringEquals": {
          "accounts.google.com:aud": "996125414232-s922bvdt21nceeh5dq1gb6av8plpj7hr.apps.googleusercontent.com",
          "accounts.google.com:sub": "SUBJECT_ID"
        }
      }
…
```

`SUBJECT_ID` can be obtained from the service account or regular account in the ID token returned from `POST https://accounts.google.com/o/oauth2/token` [decoded as JWT](https://jwt.io/).

- [ ] Include the following OAuth scopes in the connection profile
```
…
<key>Scopes</key>
<array>
    <string>openid</string>
</array>
…
```
- [ ] Set the property `s3.assumerole.rolearn` in the connection profile to the Role ARN. Set it to `s3.assumerole.rolearn=` for a prompt to enter on login.

## 🟥 AWS S3/STS + AWS IAM Identity Center [^8] 

[^8]: https://docs.aws.amazon.com/pdfs/singlesignon/latest/OIDCAPIReference/ssooidc-api.pdf

> The IAM Identity Center OIDC service currently implements only the portions of the OAuth 2.0 Device
Authorization Grant standard (https://tools.ietf.org/html/rfc8628) that are necessary to enable single
sign-on authentication with the AWS CLI. Support for other OIDC flows frequently needed for native
applications, such as Authorization Code Flow (+ PKCE), will be addressed in future releases.

Use the _[S3 (Credentials from AWS Command Line Interface)](https://docs.cyberduck.io/protocols/s3/#connecting-using-credentials-from-aws-command-line-interface)_ connection profile instead.

## 🟢 AWS S3/STS + Azure AD (OIDC)

- [x] Register OAuth Client ID in App Registrations of [Azure Portal](https://portal.azure.com/) or use the registration from Cyberduck connection profiles for [MIcrosoft OneDrive](https://github.com/iterate-ch/cyberduck/blob/master/profiles/default/Microsoft%20OneDrive.cyberduckprofile).
- [ ] Add the identity provider to AWS IAM. Obtain the thumbprint from `https://login.microsoftonline.com/<tenant>/v2.0`. The _Audience_ must be set to the OAuth Client ID like `f40bc18f-cd02-4212-b7f1-15243e4e2ad3`.
- [ ] [Configure the role and trust policy](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-amazon-web-services#configuring-the-role-and-trust-policy) in AWS IAM
- [ ] Make sure to restrict access by configuring the role and trust policy

```
…
"Condition": {
    "StringEquals": {
        "login.microsoftonline.com/<tenant>/v2.0:sub": "ge3T50cKalgsarCE6vDa34_kTb_P08X229IeC7PWqT8"
        "login.microsoftonline.com/<tenant>/v2.0:aud": "f40bc18f-cd02-4212-b7f1-15243e4e2ad3"
    }
}
…
```

- [ ] Include the following OAuth scopes in the connection profile
```
…
<key>Scopes</key>
<array>
    <string>openid</string>
    <string>offline_access</string>
</array>
…
```
- [ ] Set the property `s3.assumerole.rolearn` in the connection profile to the Role ARN. Set it to `s3.assumerole.rolearn=` for a prompt to enter on login.

### PR DESCRIPTION
**Pull request by dkocher**
Resolves #55.

### CODE REVIEW

**Comment by dkocher** [comment]
Depends on [^1] and [^2].

[^1]: iterate-ch/cyberduck#14978 
[^2]: iterate-ch/cyberduck#14972.

File: AWS S3+STS & Azure Active Directory (Azure AD) OpenID Connect.cyberduckprofile [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        </array>
+        <key>Password Configurable</key>
+        <false/>
+        <key>Username Configurable</key>
+        <false/>
+        <key>Username Placeholder</key>
+        <string>Username</string>
+        <key>Properties</key>
+        <array>
+            <!-- Require Role ARN input from user -->
```
Comment (chenkins) [inline AWS S3+STS & Azure Active Directory (Azure AD) OpenID Connect.cyberduckprofile:53]: When are users required to give this input? When creating the bookmark?

File: AWS S3+STS & Google OpenID Connect.cyberduckprofile [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        </array>
+        <key>Password Configurable</key>
+        <false/>
+        <key>Username Configurable</key>
+        <false/>
+        <key>Username Placeholder</key>
+        <string>Username</string>
+        <key>Properties</key>
+        <array>
+            <!-- Require Role ARN input from user-->
```
Comment (chenkins) [inline AWS S3+STS & Google OpenID Connect.cyberduckprofile:52]: dito.

File: AWS S3+STS & Google OpenID Connect.cyberduckprofile [hunk 3]
```diff
@@ -0,0 +1,56 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (chenkins) [inline AWS S3+STS & Google OpenID Connect.cyberduckprofile:1]: Add integration test for this protocol in `cyberduck`.

File: AWS S3+STS & Azure Active Directory (Azure AD) OpenID Connect.cyberduckprofile (hunk 1 above)
Comment (dkocher) [reply AWS S3+STS & Azure Active Directory (Azure AD) OpenID Connect.cyberduckprofile:53]: The prompt is displayed when connecting and saved in the bookmark as a property. In https://github.com/iterate-ch/cyberduck/pull/14972/files#diff-df40981f50d48245937b3cc879c98c6185d507f12496bf65f1f96a5f18130db6R121-R132

File: AWS S3+STS & Azure Active Directory (Azure AD) OpenID Connect.cyberduckprofile [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        <key>OAuth Authorization Url</key>
+        <string>https://login.microsoftonline.com/common/oauth2/v2.0/authorize</string>
+        <key>OAuth Token Url</key>
+        <string>https://login.microsoftonline.com/common/oauth2/v2.0/token</string>
+        <key>OAuth Client ID</key>
+        <string>f40bc18f-cd02-4212-b7f1-15243e4e2ad3</string>
+        <key>OAuth Client Secret</key>
+        <string/>
+        <key>OAuth Redirect Url</key>
+        <string>${oauth.handler.scheme}://oauth</string>
```
[the remark this answers is not in the document]
Comment (dkocher) [reply AWS S3+STS & Azure Active Directory (Azure AD) OpenID Connect.cyberduckprofile:39]: This is specific to the app registration possibilities and restrictions in Azure.

File: AWS S3+STS & Azure Active Directory (Azure AD) OpenID Connect.cyberduckprofile [hunk 5]
```diff
@@ -0,0 +1,57 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (dkocher) [reply AWS S3+STS & Azure Active Directory (Azure AD) OpenID Connect.cyberduckprofile:1]: Independent of this in other repository.

**Pull request merged**